// lib/lib/functional/predicates.js
"use strict";
var part = require("./partialApplication.js");
////////////////////////////////////////////////////////////////////////////////
function typeOf(s, x) {
    return typeof x === s;
}
////////////////////////////////////////////////////////////////////////////////
function instanceOf( t,x) {
    return x instanceof t;
}
////////////////////////////////////////////////////////////////////////////////
function isObject(x) {
    return typeOf("object",x) && x !== null;
}
////////////////////////////////////////////////////////////////////////////////
function isNull(x) {
    return x === null;
}
////////////////////////////////////////////////////////////////////////////////
function isArray(a) {
    return Array.isArray(a);
}
////////////////////////////////////////////////////////////////////////////////
var getProp = part.curry(function getProp(s,o) {
    return o[s];
});
var isFunc = part.curry(typeOf)('function');
var compose = require("./functors.js").compose;
var is = {
    null: isNull,
    object: isObject,
    promise:compose(getProp('then'),isFunc),
    array: isArray,
    any:(() => true),
    typeof: part.curry(typeOf),
    undefined: part.curry(typeOf)("undefined"),
    string: part.curry(typeOf)("string"),
    instanceof: part.curry(instanceOf),
    function: part.curry(typeOf)("function"),
};
////////////////////////////////////////////////////////////////////////////////
////////////////////////////////////////////////////////////////////////////////
is.maybe = require("../wrappers/object.js")
    .mask(is)
    .map(c => part.curry(function(a, f, g) {
        if (c(a)) return f(a);
        return is.function(g) ? g(a) : (a) => a;
    }))
    .o;
////////////////////////////////////////////////////////////////////////////////
module.exports = is;
